### PR TITLE
Add DMatrix usage examples to c-api-demo

### PR DESCRIPTION
**Pull request by alexanderGugel**
I had a bit of a hard time using the C API given that there aren't that many usage examples, so maybe this will be useful for people.

This PR extends the C API demo to construct a dense matrix using `XGDMatrixCreateFromMat`. I'm planning on adding additional examples for `XGDMatrixCreateFromCSREx`, `XGDMatrixCreateFromCSCEx`, but unfortunately I haven't been able to get those working yet.

The matrix being constructed is equivalent to the first line of the test data.z

### Test Plan

```
alexanders-mbp:c-api alex$ make run
cc -O3 -I../../include -I../../dmlc-core/include -I../../rabit/include -L../../lib -o c-api-demo c-api-demo.c -lxgboost
LD_LIBRARY_PATH=../../lib ./c-api-demo
[17:32:50] 6513x127 matrix with 143286 entries loaded from ../data/agaricus.txt.train
[17:32:50] 1611x127 matrix with 35442 entries loaded from ../data/agaricus.txt.test
[0]     train-error:0.014433    test-error:0.016139
[1]     train-error:0.014433    test-error:0.016139
[2]     train-error:0.014433    test-error:0.016139
[3]     train-error:0.008598    test-error:0.009932
[4]     train-error:0.001228    test-error:0.000000
[5]     train-error:0.001228    test-error:0.000000
[6]     train-error:0.001228    test-error:0.000000
[7]     train-error:0.001228    test-error:0.000000
[8]     train-error:0.001228    test-error:0.000000
[9]     train-error:0.001228    test-error:0.000000
y_pred: 0.0239 0.9544 0.0239 0.0239 0.0490 0.1056 0.9544 0.0288 0.9544 0.0242 
y_test: 0.0000 1.0000 0.0000 0.0000 0.0000 0.0000 1.0000 0.0000 1.0000 0.0000 
Dense Matrix Example (XGDMatrixCreateFromMat): 0.0239 
```

### CODE REVIEW

**Comment by trivialfis** [comment]
@alexanderGugel Would you like me to help with some other functions?

**Comment by alexanderGugel** [comment]
@trivialfis - I managed to add two other examples demonstrating how to make predictions based on sparse matrices constructed via `XGDMatrixCreateFromCSREx` / `XGDMatrixCreateFromCSCEx`.

The actual matrix being constructed is always the first row of the agaricus test set.

```
alexanders-mbp:c-api alex$ make run
cc -O3 -g -I../../include -I../../dmlc-core/include -I../../rabit/include -L../../lib -o c-api-demo c-api-demo.c -lxgboost
LD_LIBRARY_PATH=../../lib ./c-api-demo
[01:41:37] 6513x127 matrix with 143286 entries loaded from ../data/agaricus.txt.train
[01:41:37] 1611x127 matrix with 35442 entries loaded from ../data/agaricus.txt.test
[0]     train-error:0.014433    test-error:0.016139
[1]     train-error:0.014433    test-error:0.016139
[2]     train-error:0.014433    test-error:0.016139
[3]     train-error:0.008598    test-error:0.009932
[4]     train-error:0.001228    test-error:0.000000
[5]     train-error:0.001228    test-error:0.000000
[6]     train-error:0.001228    test-error:0.000000
[7]     train-error:0.001228    test-error:0.000000
[8]     train-error:0.001228    test-error:0.000000
[9]     train-error:0.001228    test-error:0.000000
y_pred: 0.0239 0.9544 0.0239 0.0239 0.0490 0.1056 0.9544 0.0288 0.9544 0.0242 
y_test: 0.0000 1.0000 0.0000 0.0000 0.0000 0.0000 1.0000 0.0000 1.0000 0.0000 
Dense Matrix Example (XGDMatrixCreateFromMat): 0.0239 
Sparse Matrix Example (XGDMatrixCreateFromCSREx): 0.0239 
Sparse Matrix Example (XGDMatrixCreateFromCSCEx): 0.0239 
```

What do you think about extending the C demo with those additional usage examples?

**Review by hcho3 — APPROVED** [review]
Would like to merge this soon, after I set up testing for the C API demo (#5910)

**Comment by hcho3** [comment]
Will revisit once #6159 is merged.

**Review by hcho3 — APPROVED** [review]
LGTM

**Pull request merged**
